### PR TITLE
4/8 - #96 subtask/encapsulate responsibilities for HttpRequest and RequestLocation

### PR DESCRIPTION
**Pull request by vdedios**
Solves #96

### CODE REVIEW

**Review by gbudau — APPROVED** [review]
Great job, thank you. I think the changes you made in this PR have made the code a lot more understandable. LGTM.

File: srcs/app/CGI/CGI.cpp [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
 									   // function somewhere common
 		 HttpResponse *response
 	) :
 	execRet_(0),
 	request_(request),
-	request_location_(&location),
+	requestConfig_(&location),
 	reqBody_(request.GetBody()),
-	arg_path_(request_location_->common.root + request.GetPath()),
+	arg_path_(requestConfig_->getRoot() + request.GetPath()),
```
Comment (Damrod) [inline srcs/app/CGI/CGI.cpp:20]: The idea of the Function naming vs variable naming (besides the underscore for private stuff) is

PascalCase

FunctionWithSomeName

VS

camelCase

variableWithSomeOtherName

Isn't it?

File: srcs/app/http/RequestConfig.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	common_(location_ ? location_->common : server_config.common),
+	path_(location_ ? &location_->path : NULL),
+	limit_except_(location_ ? &location_->limit_except : NULL),
+	root_(common_.root),
+	client_max_body_size_(common_.client_max_body_size),
+	autoindex_(common_.autoindex),
+	index_(common_.index),
+	error_pages_(common_.error_pages),
+	upload_store_(common_.upload_store),
+	return_status_(server_config.common.return_url.empty() ?
```
Comment (Damrod) [inline srcs/app/http/RequestConfig.cpp:16]: Maybe we can do a new private function for these two ternaries?
Maybe not worth it...

File: srcs/app/http/RequestConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		server_config.common.return_status :
+		common_.return_status),
+	return_url_(server_config.common.return_url.empty() ?
+		server_config.common.return_url :
+		common_.return_url),
+	cgi_assign_(common_.cgi_assign) {
+}
+
+bool RequestConfig::limits(std::string method) const {
```
Comment (Damrod) [inline srcs/app/http/RequestConfig.cpp:25]: couldn't we pass strings by const & ?

Also we could improve the naming of this function, something like

DoesLocationNotSupportMethod

(or invert the return value and name it DoesLocationSupportMethod )

File: srcs/app/http/RequestConfig.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		std::find(
+			limit_except_->begin(),
+			limit_except_->end(),
+			method) == limit_except_->end()) {
+		return true;
+	}
+	return false;
+}
+
+std::string	RequestConfig::getPath() const {
```
Comment (Damrod) [inline srcs/app/http/RequestConfig.cpp:37]: camelCase VS PascalCase, again, in most functions in this file...

File: srcs/app/http/RequestConfig.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+std::string	RequestConfig::getErrorPagePath(std::size_t errCode) const {
+	CommonConfig::ErrorPagesMap::const_iterator it =
+										error_pages_.find(errCode);
+	return it == error_pages_.end() ? "" : root_ + it->second;
+}
+
+CommonConfig::BinaryHandlerPath
+	RequestConfig::getCGIBin(std::string extension) const {
+	return cgi_assign_.find(extension)->second;
```
Comment (Damrod) [inline srcs/app/http/RequestConfig.cpp:77]: what if cgi_assign_.find.(extension) == cgi_assign_.end() ?

File: srcs/incs/RequestConfig.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		bool	hasReturnUrl() const;
+		bool	hasCGI(std::string extension) const;
+		std::string	getErrorPagePath(std::size_t errCode) const;
+		CommonConfig::BinaryHandlerPath getCGIBin(std::string extension) const;
+
+		// temporal
+		const CommonConfig	&getCommonConfig() const;
+
+	private:
+		RequestConfig &	operator=(const RequestConfig &);
```
Comment (Damrod) [inline srcs/incs/RequestConfig.hpp:34]: Why is this here? Is it implemented?

**Review by Damrod — APPROVED** [review]
Interesting, although, I believe, there is still some work to be done.

File: srcs/incs/RequestConfig.hpp (hunk 6 above)
Comment (vdedios) [reply srcs/incs/RequestConfig.hpp:34]: A lot of work in these PRs is about moving things aroung and cleaning at the end. This may be one of those left uncleaned. Thanks!

File: srcs/app/http/RequestConfig.cpp (hunk 2 above)
Comment (vdedios) [reply srcs/app/http/RequestConfig.cpp:16]: agree with you, will update it!

File: srcs/app/http/RequestConfig.cpp (hunk 3 above)
Comment (vdedios) [reply srcs/app/http/RequestConfig.cpp:25]: I know this method itself may not mean a lot. But the idea was to give context when invoked, for example:

```c++
	if (requestConfig_->limits(request.GetMethod()))
```

File: srcs/app/http/RequestConfig.cpp (hunk 5 above)
Comment (vdedios) [reply srcs/app/http/RequestConfig.cpp:77]: Totally agree, and I had a little bit of mixed feelings of this. This method is used only here, but after `IsCGI` method with already checks what you say:

```c++
std::string CGI::GetExecutable_(const std::string &extension) {
	if (requestConfig_->hasCGI(extension)) {
		return requestConfig_->getCGIBin(extension);
	}
	throw std::invalid_argument(
		"There is no CGI handler for the requested file");
}
```

Since `CGI` has still to be reworked, this was an enough good solution for me, for now.
